src/builder.cpp

```diff
@@ -0,0 +1,312 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#include "parser/specification.hpp"
+#include <builder.hpp>
+#include <charconv>
+#include <exception.hpp>
+#include <log.hpp>
+#include <parser/common.hpp>
+#include <parser/parser.hpp>
+#include <string_view>
+
+namespace ddwaf {
+
+constexpr builder::change_state operator|(builder::change_state lhs, builder::change_state rhs)
+{
+    return static_cast<builder::change_state>(
+        static_cast<std::underlying_type<builder::change_state>::type>(lhs) |
+        static_cast<std::underlying_type<builder::change_state>::type>(rhs));
+}
+
+constexpr builder::change_state operator&(builder::change_state lhs, builder::change_state rhs)
+{
+    return static_cast<builder::change_state>(
+        static_cast<std::underlying_type<builder::change_state>::type>(lhs) &
+        static_cast<std::underlying_type<builder::change_state>::type>(rhs));
+}
+
+std::shared_ptr<ruleset> builder::build(parameter object, ruleset_info &info)
+{
+    parameter::map ruleset = object;
+
+    unsigned version = parser::parse_schema_version(ruleset);
+    if (version == 1) {
+        ddwaf::ruleset rs;
+        parser::v1::parse(ruleset, info, rs, limits_);
+        return std::make_shared<ddwaf::ruleset>(std::move(rs));
+    }
+
+    if (version != 2) {
+        DDWAF_ERROR("incompatible ruleset version %u.x", version);
+        throw unsupported_version();
+    }
+
+    return build_helper(ruleset, info);
+}
+
+namespace {
+
+std::set<rule::ptr> target_to_rules(const std::vector<parser::rule_target_spec> &targets,
+    const std::unordered_map<std::string_view, rule::ptr> &rules, const rule_tag_map &rules_by_tags)
+{
+    std::set<rule::ptr> rule_targets;
```

Comment:
`rule::ptr` being a shared_ptr (this could be made more explicit btw, say using `sptr` instead of `ptr`), this method has a lot of unnecessary refcount changes. There is actually no shared ownership required here.

Reply:
I agree I need to change this to use raw pointers

Reply:
This will be a slightly big change so I'll do so for 1.8.1